### PR TITLE
Incorrect rendering of `SnapshotWidget`

### LINKED ISSUE #114398: `SnapshotWidget` renders incorrectly when enabled
Please see linked PR for reproduction

copied here:

In short, the SnapshotWidget is rendered differently when enabled vs disabled.

Reproduction:

<details>

```dart
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final binding = TestWidgetsFlutterBinding.ensureInitialized();
  testWidgets('SnapshotWidget should have same result when enabled', (tester) async {
    binding.window
      ..physicalSizeTestValue = const Size(10, 10)
      ..devicePixelRatioTestValue = 1;
    addTearDown(() => binding.window
      ..clearPhysicalSizeTestValue()
      ..clearDevicePixelRatioTestValue());

    final controller = SnapshotController(allowSnapshotting: false);
    await tester.pumpWidget(MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Container(
        color: Colors.black,
        padding: const EdgeInsets.only(right: 0.6, bottom: 0.6),
        child: SnapshotWidget(
          controller: controller,
          child: Container(
            margin: const EdgeInsets.only(right: 0.4, bottom: 0.4),
            color: Colors.blue,
          ),
        ),
      ),
    ));

    final imageWhenDisabled = await _captureImage(tester.element(find.byType(MaterialApp)));

    controller.allowSnapshotting = true;
    await tester.pump();

    final imageWhenEnabled = await _captureImage(tester.element(find.byType(MaterialApp)));
    await expectLater(imageWhenEnabled, matchesReferenceImage(imageWhenDisabled));
  });
}

Future<ui.Image> _captureImage(Element element) {
  assert(element.renderObject != null);
  RenderObject renderObject = element.renderObject!;
  while (!renderObject.isRepaintBoundary) {
    renderObject = renderObject.parent! as RenderObject;
  }
  assert(!renderObject.debugNeedsPaint);
  final OffsetLayer layer = renderObject.debugLayer! as OffsetLayer;
  return layer.toImage(renderObject.paintBounds);
}
```

</details>

If you dump the images, will see:

imageWhenDisabled
image

imageWhenEnabled
image

### PR DESCRIPTION
**Pull request by fzyzcjy**
Close https://github.com/flutter/flutter/issues/114398

### Problem description

In short, the SnapshotWidget is rendered differently when enabled vs disabled.

Reproduction:

<details>

```dart
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final binding = TestWidgetsFlutterBinding.ensureInitialized();
  testWidgets('SnapshotWidget should have same result when enabled', (tester) async {
    binding.window
      ..physicalSizeTestValue = const Size(10, 10)
      ..devicePixelRatioTestValue = 1;
    addTearDown(() => binding.window
      ..clearPhysicalSizeTestValue()
      ..clearDevicePixelRatioTestValue());

    final controller = SnapshotController(allowSnapshotting: false);
    await tester.pumpWidget(MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Container(
        color: Colors.black,
        padding: const EdgeInsets.only(right: 0.6, bottom: 0.6),
        child: SnapshotWidget(
          controller: controller,
          child: Container(
            margin: const EdgeInsets.only(right: 0.4, bottom: 0.4),
            color: Colors.blue,
          ),
        ),
      ),
    ));

    final imageWhenDisabled = await _captureImage(tester.element(find.byType(MaterialApp)));

    controller.allowSnapshotting = true;
    await tester.pump();

    final imageWhenEnabled = await _captureImage(tester.element(find.byType(MaterialApp)));
    await expectLater(imageWhenEnabled, matchesReferenceImage(imageWhenDisabled));
  });
}

Future<ui.Image> _captureImage(Element element) {
  assert(element.renderObject != null);
  RenderObject renderObject = element.renderObject!;
  while (!renderObject.isRepaintBoundary) {
    renderObject = renderObject.parent! as RenderObject;
  }
  assert(!renderObject.debugNeedsPaint);
  final OffsetLayer layer = renderObject.debugLayer! as OffsetLayer;
  return layer.toImage(renderObject.paintBounds);
}
```

</details>

If you dump the images, will see:

imageWhenDisabled
image

imageWhenEnabled
image

### How PR solves it

It seems that this is caused by integer rounding. For example, suppose our SnapshotWidget (and thus its Layer) is 9.4 pixels, then after toImageSync will get a 10x10 (or 9x9?) ui.Image instead of a 9.4x9.4 image - since image size must be integer. Then, next time we paint this ui.Image, the original code will paint the 10x10 rectangle area into the 9.4x9.4 Canvas, thus causing resizing of the content. The PR will remember the real size is 9.4 instead of 10, so next time when painting the ui.Image we will paint the 9.4x9.4 rectangle area into the 9.4x9.4 canvas, so no resize of content.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jonahwilliams — CHANGES_REQUESTED** [review]
Please write descriptions for the issues you file and for an explanation on the PRs that you have opened.

**Comment by fzyzcjy** [comment]
@jonahwilliams done

File: packages/flutter/test/widgets/snapshot_widget_test.dart [hunk 1]
```diff
@@ -261,6 +263,39 @@ void main() {
     expect(tester.takeException(), isNull);
     expect(tester.layers.last, isA<PlatformViewLayer>());
   }, skip: kIsWeb); // TODO(jonahwilliams): https://github.com/flutter/flutter/issues/106689
+
+  testWidgets('SnapshotWidget should have same result when enabled', (WidgetTester tester) async {
+    binding.window
```
Comment (jonahwilliams) [inline packages/flutter/test/widgets/snapshot_widget_test.dart:268]: The tester has a binding too

File: packages/flutter/test/widgets/snapshot_widget_test.dart [hunk 2]
```diff
@@ -261,6 +263,39 @@ void main() {
...
+      ..physicalSizeTestValue = const Size(10, 10)
+      ..devicePixelRatioTestValue = 1;
+    addTearDown(() => binding.window
```
Comment (jonahwilliams) [inline packages/flutter/test/widgets/snapshot_widget_test.dart:271]: I think we normally use `try { } finally { }` for these. Maybe we should use addTearDown if its easier

File: packages/flutter/test/widgets/snapshot_widget_test.dart [hunk 3]
```diff
@@ -261,6 +263,39 @@ void main() {
...
+      ..clearPhysicalSizeTestValue()
+      ..clearDevicePixelRatioTestValue());
+
+    final SnapshotController controller = SnapshotController();
+    await tester.pumpWidget(MaterialApp(
```
Comment (jonahwilliams) [inline packages/flutter/test/widgets/snapshot_widget_test.dart:276]: Add a repaint boundary here and then use the normal method for screenshot testing?

File: packages/flutter/lib/src/widgets/snapshot_widget.dart [hunk 4]
```diff
@@ -368,7 +377,7 @@ abstract class SnapshotPainter extends ChangeNotifier  {
   /// }
   /// ```
   /// {@end-tool}
-  void paintSnapshot(PaintingContext context, Offset offset, Size size, ui.Image image, double pixelRatio);
+  void paintSnapshot(PaintingContext context, Offset offset, Size size, ui.Image image, Size imageSize, double pixelRatio);
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/snapshot_widget.dart:380]: Nit on the naming, its not really the image size. Maybe call it the srcSize? Not sure exactly.

Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/snapshot_widget.dart:380]: Please update the docs to include the new parameter and how to use it.

**Review by jonahwilliams — COMMENTED** [review]
Thanks for tracking this down, this is a tricky one!

File: packages/flutter/test/widgets/snapshot_widget_test.dart (hunk 2 above)
Comment (fzyzcjy) [reply packages/flutter/test/widgets/snapshot_widget_test.dart:271]: Depends on you, if you like I can refactor to try-finally

In addition, I personally have a utility function internally, so I just write

```
binding.window.setUpTearDown(physicalSizeTestValue: ..., )
```

and it auto setup and teardown

if you like I can PR that in a separate PR

File: packages/flutter/test/widgets/snapshot_widget_test.dart (hunk 1 above)
Comment (fzyzcjy) [reply packages/flutter/test/widgets/snapshot_widget_test.dart:268]: done

File: packages/flutter/lib/src/widgets/snapshot_widget.dart (hunk 4 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/snapshot_widget.dart:380]: done

Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/snapshot_widget.dart]: done

File: packages/flutter/test/widgets/snapshot_widget_test.dart (hunk 3 above)
Comment (fzyzcjy) [reply packages/flutter/test/widgets/snapshot_widget_test.dart]: done

File: packages/flutter/test/widgets/snapshot_widget_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/test/widgets/snapshot_widget_test.dart]: I think the addTearDown is fine!

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Comment by jonahwilliams** [comment]
FYI @dnfield / @goderbauer as secondary review

File: packages/flutter/lib/src/widgets/snapshot_widget.dart [hunk 5]
```diff
@@ -356,19 +365,21 @@ abstract class SnapshotPainter extends ChangeNotifier  {
...
+  ///   final Rect src = Rect.fromLTWH(0, 0, srcSize.width, srcSize.height);
   ///   final Rect dst = Rect.fromLTWH(offset.dx, offset.dy, size.width, size.height);
   ///   final Paint paint = Paint()
   ///     ..filterQuality = FilterQuality.low;
   ///   context.canvas.drawImageRect(image, src, dst, paint);
   /// }
   /// ```
   /// {@end-tool}
-  void paintSnapshot(PaintingContext context, Offset offset, Size size, ui.Image image, double pixelRatio);
+  void paintSnapshot(PaintingContext context, Offset offset, Size size, ui.Image image, Size srcSize, double pixelRatio);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/snapshot_widget.dart:382]: srcSize: In accordance with our style guide, please avoid abbreviations, see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-abbreviations

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/snapshot_widget.dart (hunk 5 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/snapshot_widget.dart:382]: Then sourceSize?

**Comment by jonahwilliams** [comment]
@fzyzcjy could you update to ToT so we can land this?

**Comment by fzyzcjy** [comment]
@jonahwilliams sure, done (waiting for ci now)

**Pull request merged**